### ISSUE TITLE
NumPyroConverter makes too many assumptions about the model

### ISSUE THREAD

**Issue by torfjelde** [arviz-devs/arviz#1704]
The following code is not always runnable:

https://github.com/arviz-devs/arviz/blob/7ebedd2d1494f07513bc7b9ea21962bc11399a94/arviz/data/io_numpyro.py#L126-L133

This will fail for a model involving `ImproperUniform`, as sampling for it is not defined.

Would it be possible to add a keyword argument that specifies whether this automatic extraction of observations should be attempt or not, e.g. `extract_observations` or something? At the moment it's stopping me from being able to use arviz with my numpyro-model (unless I just comment out the lines above) :confused:

**Comment by OriolAbril** [comment]
I think adding an extra argument to skip this sounds fine, as would probably be adding a try except or even both. cc @fehiepsi

**Comment by fehiepsi** [comment]
Because this just requires a valid trace, I believe we can just use some sort of init_to_uniform strategy. I'll push a PR shortly.

**Comment by torfjelde** [comment]
> Because this just requires a valid trace, I believe we can just use some sort of init_to_uniform strategy. I'll push a PR shortly.

True, but even so, as a user, I'd prefer not always inferring this, i.e. at least allow me to turn it off. E.g. the observations might be not be present because I'm "observing" using a factor, in which case the automatically inferred observations can be misleading.

**Comment by fehiepsi** [comment]
That's a good point! I think you can set `observations = None` or delete some of its keys if you want. It will give you some flexibility in extracting `log_likelihood` or something like that.

**Comment by torfjelde** [comment]
How about adding a kwarg `observations=None` and then do `if observations is None and self.model is not None` or something along those lines? This way you can choose not to extract the observations by setting `observations={}` or any empty collection.

**Comment by fehiepsi** [comment]
> How about adding a kwarg observations=None and then do if observations is None and self.model is not None or something along those lines?

Sounds reasonable to me. Do you want to submit a PR? :)

What do you think @OriolAbril? We can filter out `factor` variables (those variables contain log probabilities, which plays the role of `target += ...` in [Stan](https://mc-stan.org/docs/2_26/stan-users-guide/examples.html#triangle-distribution)) but I'm not sure if arviz's log_likelihood will include those terms (in Stan, PyMC3,...?) In my opinion, those extra log probabilities play the role of log-likelihoods.

**Comment by OriolAbril** [comment]
> > How about adding a kwarg observations=None and then do if observations is None and self.model is not None or something along those lines?
>
> Sounds reasonable to me. Do you want to submit a PR? :)

Yeah, it does sound great, some other converters already have an argument for the observed data. I think that for coherence with those converters, the argument should be `oberved_data`, then for what it should take, I think a boolean to keep or disable the current behaviour makes sense, not sure if taking a dict of "extra" variables to add or override in the `observed_data` would make sense too. 

> What do you think @OriolAbril? We can filter out factor variables (those variables contain log probabilities, which plays the role of target += ... in Stan) but I'm not sure if arviz's log_likelihood will include those terms (in Stan, PyMC3,...?) In my opinion, those extra log probabilities play the role of log-likelihoods.

In Stan the generation of pointwise log likelihood values is completely manual, ArviZ only takes strings that indicate the variable names storing log lik data. For pymc, pm.Potential terms which are the ones that play this role are ignored by the auto log likelihood retrieval function of the converter. And I think it should be kept this way, I don't think there can be any guarantees when converting potentials or factors to _pointwise_ log likelihood values. They should be added afterwards by the user to ensure they are what they are supposed to be. We have an example of adding log likelihood directly to inferencedata object at https://arviz-devs.github.io/arviz/user_guide/numpyro_refitting_xr_lik.html (it's a toy case where we delete it and add it again but should be enough for illustration purposes)

**Comment by fehiepsi** [comment]
>  we're using ImproperUniform together with a factor in the prior to make the implementation more efficient.

Oh, that's a good point! I forgot that `factor` can be used this way. ;)
